File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        if len(nums) == 1: return nums

        valToFreq = {}
        freqToVal = {}
        
        for i in range(0, len(nums)):
            if valToFreq.get(nums[i]) == None:
                valToFreq[nums[i]] = 1

                if freqToVal.get(1) == None:
                    freqToVal[1] = {nums[i]}
                else:
                    freqToVal[1].add(nums[i])
            
            else:
                valToFreq[nums[i]] += 1
                if freqToVal.get(valToFreq[nums[i]]) == None:
                    freqToVal[valToFreq[nums[i]]] = {nums[i]}
                    freqToVal[valToFreq[nums[i]]-1].remove(nums[i])
                else:
                    freqToVal[valToFreq[nums[i]]].add(nums[i])
                    freqToVal[valToFreq[nums[i]]-1].remove(nums[i])
        output = []
        j = 0
        Max = max(freqToVal.keys())

        while len(output) < k:
            output += freqToVal[Max-j]
            j +=1

        return output
        
        

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count = {};
        vector<vector<int>> freq(nums.size() + 1);

        for (int n : nums) {
            if (count.find(n) != count.end()) ++count[n];
            else count[n] = 1;
        }

        for (auto const& entry : count) {
            freq[entry.second].push_back(entry.first);
        };

        vector<int> res = {};
        for (int j = 1; j < freq.size(); ++j) {
            for (int n = 0; n < freq[freq.size() - j].size(); ++n) {
                res.push_back(freq[freq.size() - j][n]);
                if (res.size() == k) {
                    return res;
                };
            };
        };
        return res;
    };
};